### PR TITLE
feat(extensions): add `user skipped` task hook

### PR DESCRIPTION
**Pull request by vikaspotluri123**
refs #308, #1115

I'm not sure if this is the best way to do this, but it's the most pluggable thing that came to mind. I haven't fully tested this or ensured that changing the process manager the way I did will actually work 😄 It's best to look at 28a85b8 individually since it's the only logical change.

@acburdine can you please take a cursory look at this?

### CODE REVIEW

File: extensions/systemd/index.js [hunk 1]
```diff
@@ -22,6 +22,9 @@ class SystemdExtension extends Extension {
                 }
 
                 return false;
+            },
+            userSkip: ({instance}) => {
```
Comment (acburdine) [inline extensions/systemd/index.js:26]: I think this approach makes sense - though I wonder if it might be better named, like `onSkip` or something 🤔

Comment (vikaspotluri123) [reply extensions/systemd/index.js:26]: My issue with `onSkip` is it's ambiguous as to what skipped (auto or user) - thoughts on `onUserSkip`?

Comment (acburdine) [reply extensions/systemd/index.js:26]: yeah that sounds good 👍

File: lib/commands/setup.js [hunk 2]
```diff
@@ -141,6 +141,10 @@ class SetupCommand extends Command {
 
                 const confirmed = await this.ui.confirm(`Do you wish to set up ${name}?`, true);
                 if (!confirmed) {
+                    if (step.onUserSkip) {
+                        step.onUserSkip(ctx);
```
Comment (acburdine) [inline lib/commands/setup.js:145]: do we want to add an await here just to support any promise-based skips?

Comment (vikaspotluri123) [reply lib/commands/setup.js:145]: Sure

File: extensions/systemd/index.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ class SystemdExtension extends Extension {
                 }
 
                 return false;
+            },
+            onUserSkip: ({instance}) => {
+                instance.config.set('process', 'local').save();
```
Comment (acburdine) [inline extensions/systemd/index.js:27]: last thing - do we want to print out a log message saying that we're reverting back to local process manager?

Comment (vikaspotluri123) [reply extensions/systemd/index.js:27]: Ooh yeah, good idea - `Using "local" process manager` (yellow)?

Comment (acburdine) [reply extensions/systemd/index.js:27]: maybe "Systemd setup skipped, reverting to local process manager", yellow sounds good color-wise though

**Review by acburdine — APPROVED** [review]

**Comment by acburdine** [comment]
@vikaspotluri123 looks like the log isn't being supplied for the test

**Comment by acburdine** [comment]
also had another thought, would it make sense to add a check to the start of systemd setup to switch the process to systemd if it's not set to that already?

thinking if someone skips the check the first time around, if they run `ghost setup systemd` later we'd likely want to change their process manager back to systemd.

File: extensions/systemd/index.js [hunk 4]
```diff
@@ -22,18 +22,28 @@ class SystemdExtension extends Extension {
...
-    _setup({instance}, task) {
+    _setup({instance, ui}, task) {
         const uid = getUid(instance.dir);
 
         // getUid returns either the uid or null
         if (!uid) {
             return task.skip('The "ghost" user has not been created, try running `ghost setup linux-user` first');
         }
 
+        if (instance.config.get('process') !== 'systemd') {
```
Comment (acburdine) [inline extensions/systemd/index.js:41]: we might need to change the enable function as well, since it checks if process is systemd

Comment (vikaspotluri123) [reply extensions/systemd/index.js:41]: What could we do for that? Just `!argv.local`? Or `!argv.local || argv.force` 🤔

Comment (acburdine) [reply extensions/systemd/index.js:41]: maybe check if `argv.stages` contains systemd? https://github.com/TryGhost/Ghost-CLI/blob/master/lib/commands/setup.js#L127-L129

Comment (vikaspotluri123) [reply extensions/systemd/index.js:41]: Can't suggest a patch, but were you thinking this?

```js
enabled: ({instance, argv}) => !argv.local && (instance.config.get('process') === 'systemd' || argv.stages.includes('systemd')),
```

Or completely getting rid of the `instance.config` check?

Comment (acburdine) [reply extensions/systemd/index.js:41]: probably
```js
enabled: ({instance, argv}) => !argv.local && (instance.config.get('process') === 'systemd' || (argv.stages && argv.stages.includes('systemd'))),
```

Comment (acburdine) [reply extensions/systemd/index.js:41]: sidenote, can't wait until optional chaining is available in node LTS 😂

Comment (vikaspotluri123) [reply extensions/systemd/index.js:41]: If we switch to typescript we could do it today 😛

**Pull request merged**